### PR TITLE
[FIXED JENKINS-28121] Need to mkdirs before clone

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-28121](https://issues.jenkins-ci.org/browse/JENKINS-28121)

Picks up tests from https://github.com/jenkinsci/workflow-scm-step-plugin/pull/4 and https://github.com/jenkinsci/workflow-multibranch-plugin/pull/17, and makes them pass when `CACHE_LOCAL_REPOS` is deleted.

@reviewbybees

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -13,8 +13,7 @@
     <url>http://wiki.jenkins-ci.org/display/JENKINS/Mercurial+Plugin</url>
     
     <properties>
-      <jenkins.version>1.580.1</jenkins.version>
-      <java.level>6</java.level>
+      <jenkins.version>1.642.3</jenkins.version>
```
Comment (jglick) [inline pom.xml]: Need a new baseline to be able to run tests against reasonably modern versions of Pipeline; the 1.4.x line compatible with 1.580.x did not even include `workflow-multibranch`, which most naturally reproduces this error.

File: src/main/java/hudson/plugins/mercurial/MercurialSCM.java [hunk 2]
```diff
@@ -809,6 +809,7 @@ private void clone(Run<?, ?> build, Launcher launcher, FilePath repository, Node
             args.add(getSource(env));
         }
         args.add(repository.getRemote());
+        repository.mkdirs();
```
Comment (jglick) [inline src/main/java/hudson/plugins/mercurial/MercurialSCM.java]: The actual fix.

File: src/main/java/hudson/plugins/mercurial/MercurialSCM.java [hunk 3]
```diff
@@ -932,12 +933,8 @@ private boolean causedByMissingHg(IOException e) {
         return message != null && message.startsWith("Cannot run program") && message.endsWith("No such file or directory");
     }
 
-    static boolean CACHE_LOCAL_REPOS = false;
```
Comment (abayer) [inline src/main/java/hudson/plugins/mercurial/MercurialSCM.java]: So what was the point of this such that we're getting rid of it now?

Comment (jglick) [inline src/main/java/hudson/plugins/mercurial/MercurialSCM.java]: It was a little faster. But for a situation unlikely to arise in production.

**Comment by abayer** [comment]
:bee:

**Pull request merged**